expression/builtin_regexp.go

```diff
@@ -0,0 +1,1414 @@
+
+package expression
+
+import (
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"unicode/utf8"
+
+	"github.com/pingcap/tidb/parser/charset"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/stringutil"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+type empty struct{}
+
+const patternIdx = 1
+const RegexpLikeMatchTypeIdx = 2
+const RegexpSubstrMatchTypeIdx = 4
+const RegexpInstrMatchTypeIdx = 5
+const RegexpReplaceMatchTypeIdx = 5
+
+// Valid flags in match type
+const (
+	flag_i = "i"
+	flag_c = "c"
+	flag_m = "m"
+	flag_s = "s"
+)
+
+const (
+	invalidMatchType    = "Invalid match type"
+	invalidIndex        = "Index out of bounds in regular expression search"
+	invalidReturnOption = "Incorrect arguments to regexp_instr: return_option must be 1 or 0"
+)
+
+var validMatchType = map[string]empty{
+	flag_i: {}, // Case-insensitive matching
+	flag_c: {}, // Case-sensitive matching
+	flag_m: {}, // Multiple-line mode
+	flag_s: {}, // The . character matches line terminators
+}
+
+type regexpBaseFuncSig struct {
+	baseBuiltinFunc
+	regexpMemorizedSig
+	once sync.Once
+}
+
+func (re *regexpBaseFuncSig) isBinCollation() bool {
+	return re.collation == charset.CollationBin
+}
+
+func (re *regexpBaseFuncSig) clone(from *regexpBaseFuncSig) {
+	if from.memorizedRegexp != nil {
+		re.memorizedRegexp = from.memorizedRegexp.Copy()
+	}
+	re.memorizedErr = from.memorizedErr
+	re.cloneFrom(&from.baseBuiltinFunc)
+}
+
+// If characters specifying contradictory options are specified
+// within match_type, the rightmost one takes precedence.
+func (re *regexpBaseFuncSig) getMatchType(bf *baseBuiltinFunc, userInputMatchType string) (string, error) {
+	flag := ""
+	matchTypeSet := make(map[string]empty)
+
+	if collate.IsCICollation(bf.collation) {
+		matchTypeSet[flag_i] = empty{}
+	}
+
+	for _, val := range userInputMatchType {
+		c := string(val)
+
+		// Check validation of the flag
+		_, err := validMatchType[c]
+		if err == false {
+			return "", ErrRegexp.GenWithStackByArgs(invalidMatchType)
+		}
+
+		if c == flag_c {
+			// re2 is case-sensitive by default, so we only need to delete 'i' flag
+			// to enable the case-sensitive for the regexp
+			delete(matchTypeSet, flag_i)
+			continue
+		}
+
+		matchTypeSet[c] = empty{} // add this flag
+	}
+
+	// generate flag
+	for key := range matchTypeSet {
+		flag += key
+	}
+
+	return flag, nil
+}
+
+// To get a unified compile interface in initMemoizedRegexp, we need to process many things in genCompile
+func (re *regexpBaseFuncSig) genCompile(matchType string) (func(string) (*regexp.Regexp, error), error) {
+	matchType, err := re.getMatchType(&re.baseBuiltinFunc, matchType)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(pat string) (*regexp.Regexp, error) {
+		return regexp.Compile(fmt.Sprintf("(?%s)%s", matchType, pat))
```

Comment:
If `matchType` is empty string, we don't have to customize the compile function?

Reply:
> If `matchType` is empty string, we don't have to customize the compile function?

Nothing will happen when `matchType` is empty.